pkg/util/virtualmachineinstance/virtualmachineinstance.go

```diff
@@ -53,13 +53,44 @@ func GetAllNonLiveMigratableVMINames(vmis []*kubevirtv1.VirtualMachineInstance,
 		// container-disk or cdrom device
 		if VMContainsCDRomOrContainerDisk(vmi) {
 			nonLiveMigratableVMINames = append(nonLiveMigratableVMINames, vmiNamespacedName)
+			logrus.Infof("%s considered non-live migratable due to CD-ROM or container disk", vmiNamespacedName)
 			continue
 		}
 	}
 
 	return nonLiveMigratableVMINames, nil
 }
 
+func ValidateVMMigratable(vmi *kubevirtv1.VirtualMachineInstance) error {
+	vmiNamespacedName := fmt.Sprintf("%s/%s", vmi.Namespace, vmi.Name)
+
+	if !vmi.IsRunning() {
+		return fmt.Errorf("VM %s considered non-live migratable due to running", vmiNamespacedName)
+	}
+
+	// The VM is already in migrating state
+	if vmi.Annotations[util.AnnotationMigrationUID] != "" {
+		return fmt.Errorf("VM %s considered non-live migratable due to migration state", vmiNamespacedName)
+	}
+
+	// Node selectors
+	if vmi.Spec.NodeSelector != nil && vmi.Spec.NodeSelector[corev1.LabelHostname] != "" {
+		return fmt.Errorf("VM %s considered non-live migratable due to node selectors", vmiNamespacedName)
+	}
+
+	// PCIe devices
+	if vmi.Spec.Domain.Devices.HostDevices != nil {
```

Comment:
can we please also add a check for `Devices.GPUs` as that is being by vGPU feature

Reply:
BTW, the condition `len(vmi.Spec.Domain.Devices.HostDevices) != 0` is better. We don't need to check `nil` anymore.